lib/ui/painting/immutable_buffer.cc

```diff
@@ -49,6 +53,56 @@ void ImmutableBuffer::init(Dart_NativeArguments args) {
   tonic::DartInvoke(callback_handle, {Dart_TypeVoid()});
 }
 
+void ImmutableBuffer::initFromAsset(Dart_NativeArguments args) {
+  UIDartState::ThrowIfUIOperationsProhibited();
+  Dart_Handle callback_handle = Dart_GetNativeArgument(args, 2);
+  if (!Dart_IsClosure(callback_handle)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Callback must be a function"));
+    return;
+  }
+  Dart_Handle asset_name_handle = Dart_GetNativeArgument(args, 1);
+  uint8_t* chars = nullptr;
+  intptr_t asset_length = 0;
+  Dart_Handle result =
+      Dart_StringToUTF8(asset_name_handle, &chars, &asset_length);
+  if (Dart_IsError(result)) {
+    Dart_SetReturnValue(args, tonic::ToDart("Asset must be valid UTF8"));
+    return;
+  }
+  Dart_Handle immutable_buffer = Dart_GetNativeArgument(args, 0);
+
+  std::string asset_name = std::string{reinterpret_cast<const char*>(chars),
+                                       static_cast<size_t>(asset_length)};
+
+  std::shared_ptr<AssetManager> asset_manager = UIDartState::Current()
+                                                    ->platform_configuration()
+                                                    ->client()
+                                                    ->GetAssetManager();
+  std::unique_ptr<fml::Mapping> data = asset_manager->GetAsMapping(asset_name);
+  if (data == nullptr) {
+    Dart_SetReturnValue(args, tonic::ToDart("Asset not found"));
+    return;
+  }
+
+  auto size = data->GetSize();
+  const void* bytes = static_cast<const void*>(data->GetMapping());
+
+#if FML_OS_ANDROID
+  // fml::Mapping backed by android assets are not thread safe.
+  auto sk_data = MakeSkDataWithCopy(bytes, size);
+#else
+  SkData::ReleaseProc proc = [](const void* ptr, void* context) {
+    delete static_cast<fml::Mapping*>(context);
+  };
+  void* peer = reinterpret_cast<void*>(data.release());
+  auto sk_data = SkData::MakeWithProc(bytes, size, proc, peer);
+#endif  // FML_OS_ANDROID
```

Comment:
I don't think we can assume this is thread safe on any platform - we should just default to making copies everywhere, since the `fml::Mapping` interface doesn't declare itself as thread safe.

Reply:
Fair

Reply:
Done